### PR TITLE
Updates to README.md to help with readability

### PR DESCRIPTION
**Pull request by gilcrest**
Hey - I've recently decided to switch to zerolog and wanted to try and help by adding some organization to the README.md file.  I love the logger so far!  As I'm teaching myself the library and learning the APIs, it seems like there's some room for improvement in the instructions for the layman (that would be me)...  I have more edits I'd like to make as I go, but wanted to see if you considered this helpful before I spent more time on it!  Thanks for the logger and your consideration!

You can see the changes fully realized at my fork, which may be easier to read than all the code merge: [https://github.com/gilcrest/zerolog](https://github.com/gilcrest/zerolog)

Dan

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+## Getting Started
+### Simple Logging
+For simple logging, import the global logger package **github.co/rs/zerolog/log**
+```go
+package main
 
-### A global logger can be use for simple logging
+import (
+	"github.com/rs/zerolog/log"
+)
```
Comment (rs) [inline README.md]: You might make this one-line to reduce the size of the example and focus on the important stuff.

File: README.md [hunk 2]
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
-NOTE: To import the global logger, import the `log` subpackage `github.com/rs/zerolog/log`.
+**zerolog** allows for logging at the following levels (from highest to lowest):
+- panic (PanicLevel, 5)
+- fatal (FatalLevel, 4)
+- error (ErrorLevel, 3)
+- warn (WarnLevel, 2)
+- info (InfoLevel, 1)
+- debug (DebugLevel, 0)
 
+You can set the Global logging level to any of these options using the SetGlobalLevel function in the zerolog package, passing in one of the given constants above, e.g. zerolog.InfoLevel would be the "info" level.  Whichever level is chosen, all logs with a level greater than or equal to that level will be written. To turn off logging entirely, pass the Disabled constant.
```
Comment (rs) [inline README.md]: Could you please backquote "code" like (`SetGlobalLevel` and `zerolog.InfoLevel`) in the text?

File: README.md [hunk 3]
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+package main
+
+import (
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+func main() {
+	//zerolog.SetGlobalLevel(zerolog.InfoLevel) // Run 1
+	zerolog.SetGlobalLevel(zerolog.DebugLevel) // Run 2
```
Comment (rs) [inline README.md]: Would be nicer with the use of a `-d` flag instead.

File: README.md [hunk 4]
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
-## Usage
-
+## Installation
 ```go
-import "github.com/rs/zerolog/log"
+go get -u github.com/rs/zerolog/log
 ```
+## Getting Started
+### Simple Logging
+For simple logging, import the global logger package **github.co/rs/zerolog/log**
```
[the remark this answers is not in the document]
Comment (gilcrest) [reply README.md]: Yup, will do

File: README.md (hunk 1 above)
Comment (gilcrest) [reply README.md]: No problem

File: README.md [hunk 5]
```diff
@@ -24,43 +24,120 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
-### A global logger can be use for simple logging
+import (
+	"github.com/rs/zerolog/log"
+)
 
-```go
-log.Print("hello world")
+func main() {
+	// I like UNIX Time - if you set zerolog.TimeFieldFormat to an
+	// empty string, logs will write with UNIX time
```
[the remark this answers is not in the document]
Comment (gilcrest) [reply README.md]: Good to know!  I'll change this slightly to reflect that as well...

File: README.md (hunk 2 above)
Comment (gilcrest) [reply README.md]: Yes, will do

File: README.md (hunk 3 above)
Comment (gilcrest) [reply README.md]: OK, this is a good excuse to teach myself flags...  Give me a little time on this one

**Comment by rs** [comment]
Awesome, I'll merge where you're done.

What about the `log_example_test.go`?

**Comment by gilcrest** [comment]
Oh right, I'll add the examples to that as well

**Comment by gilcrest** [comment]
Hrm, so I started to add the examples from the README into `log_example_test.go`, but by nature, they all log with the current time, which is always variable and cannot be predicted in a static output...  What do you think if I add something like this to the package - it will pass the test as there's no "Output"...  Unless you have a better idea?

```go
package zerolog_test

import (
	"errors"
	stdlog "log"
	"os"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
)

// The below is a simple logging example using the log package.
// Note that "Output" is renamed to "Produces" as otherwise, this
// test will not pass. Log.Print has time in its Output, which
// cannot be predicted in a static Output
func ExampleSimple() {
	// UNIX Time is faster and smaller than most timestamps
	// If you set zerolog.TimeFieldFormat to an empty string,
	// logs will write with UNIX time
	zerolog.TimeFieldFormat = ""

	log.Print("hello world")

	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}

}
```

**Comment by gilcrest** [comment]
Also, one thing - having to add the `zerolog.TimeFieldFormat = ""` is obviously adding length to these examples and an extra import...  When I installed the package, by default I was getting "normal" timestamp ("time":"2018-01-17T10:46:18-05:00") - I can shorten all these if you're ok with displaying this type of timestamp in the various examples, but you had been using UNIX time in all your examples, so I stuck with that...  Might it even be better to have UNIX time as the default though?

**Comment by rs** [comment]
In test mode we could make the timestamp static.

Use `zerolog.TimeFieldFormat = ""` would be the obvious choice but I don't want to break people log pipeline with such change.

**Comment by gilcrest** [comment]
I added 3 of the 4 examples -- I put in the expected `Output:` - unfortunately, I'm not sure what you mean by "In test mode we could make the timestamp static."  I removed Output for now in order to get them in there and the tests don't fail.  I tried a few different things locally, but at a bit of a loss - sorry - learning Go n00b-ness.  Also, for the example where you say "Would be nicer with the use of a -d flag instead."  I have no idea what a -d flag is... I searched through the flags doc a bunch of blogs, etc but no mention of a -d flag... If you can help point me in the right direction, I should be able to figure it out...

**Comment by rs** [comment]
Basically you isolate the part that set the timestamp to "now" into a global function variable (like `timeNowFunc`) and in the _test.go's `init` function, you override this function to return a fixed time. This way, all tests will get a constant timestamp.

By `-d` flag, I mean creating a `-debug` or `-d` flag that would change the log level to debug (the default being set to info for instance. This would let you modulate the log level with a cli parameter instead of having to recompile with different commented lines.

**Comment by gilcrest** [comment]
ok, super helpful comments, this is great - I'm learning - thank you - getting there... sorry for all the back and forth...  I have just checked in the changes for the Global Log Level section using flags - it's slightly different than what you were saying - can you let me know if you're ok with the way I did it?

File: README.md [hunk 6]
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+package main
+
+import (
+	"flag"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+var lvl string
```
Comment (rs) [inline README.md]: No need for global variable if you move the flag into the main.

File: README.md [hunk 7]
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+
+// Declare "level" flag and bind it to lvl variable
+func init() {
+	flag.StringVar(&lvl, "level", "error", "level is the logging level (debug, info, warn, etc.")
```
Comment (rs) [inline README.md]: You should move this to the main. It's not a great pattern to have your flag as globals.

You may also use this notation:

```go
lvl := flag.String("level", …)
```

I would stick with a `debug` flag to simplify the example, even if you're right, this approach is more realistic.

File: README.md [hunk 8]
```diff
@@ -24,43 +24,161 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+
+func main() {
+
+	flag.Parse()
+	zerolog.TimeFieldFormat = ""
+
+	// Set the logging level based on the command line flag
+	zerolog.SetGlobalLevel(level(lvl))
+
+	log.Debug().Msg("filtered out message")
```
Comment (rs) [inline README.md]: The messages are no longer relevant.

File: log_example_test.go [hunk 9]
```diff
@@ -394,3 +421,19 @@ func ExampleContext_Durs() {
...
+	zerolog.TimeFieldFormat = ""
+
+	log.Fatal().
+		Err(err).
+		Str("service", service).
+		Msgf("Cannot start %s", service)
+
+	// Produces: {"time":1516133263,"level":"fatal","error":"A repo man spends his life getting into tense situations","service":"myservice","message":"Cannot start myservice"}
+	//         exit status 1
+
```
Comment (rs) [inline log_example_test.go]: Unneeded `exit status 1` and extra blank line.

File: log_example_test.go [hunk 10]
```diff
@@ -7,8 +7,35 @@ import (
...
+// using the log package.  You'll note the below function ExampleNew is similar,
+// but has a different log output as it's getting an instance of
+// the logger from the zerolog package instead of using the log package
+func ExampleSimpleLevelLog() {
+	zerolog.TimeFieldFormat = ""
+
+	log.Info().Msg("hello world")
+
+	// Produces: {"time":1516227604,"level":"info","message":"hello world"}
+
```
Comment (rs) [inline log_example_test.go]: Extra blank line.

File: log_example_test.go [hunk 11]
```diff
@@ -7,8 +7,35 @@ import (
...
+func ExampleSimple() {
+	// UNIX Time is faster and smaller than most timestamps
+	// If you set zerolog.TimeFieldFormat to an empty string,
+	// logs will write with UNIX time
+	zerolog.TimeFieldFormat = ""
+
+	log.Print("hello world")
+
+	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}
+
```
Comment (rs) [inline log_example_test.go]: Extra blank line.

File: README.md (hunk 6 above)
Comment (gilcrest) [reply README.md]: ok got it - thx

File: README.md (hunk 7 above)
Comment (gilcrest) [reply README.md]: Understood - thanks again!

File: README.md (hunk 8 above)
Comment (gilcrest) [reply README.md]: Updated messages to hopefully be clearer...

File: log_example_test.go (hunk 11 above)
Comment (gilcrest) [reply log_example_test.go]: Gone

File: log_example_test.go (hunk 10 above)
Comment (gilcrest) [reply log_example_test.go]: Purged

File: log_example_test.go (hunk 9 above)
Comment (gilcrest) [reply log_example_test.go]: dropped

File: README.md [hunk 12]
```diff
@@ -24,43 +24,150 @@ To keep the code base and the API simple, zerolog focuses on JSON logging only.
...
+	// Determine log level based on command line flag
+	// For this example, we set level to Info if Debug flag is not present
+	var lvl zerolog.Level
+	if *debug {
+		lvl = zerolog.DebugLevel
+	} else {
+		lvl = zerolog.InfoLevel
+	}
+
+	zerolog.SetGlobalLevel(lvl)
```
Comment (rs) [inline README.md]: Cosmetic comment: I would have written this that way to better show what's default:

```go
zerolog.SetGlobalLevel(zerolog.InfoLevel)
if *debug {
    zerolog.SetGlobalLevel(zerolog.DebugLevel)
}
```

File: log_example_test.go [hunk 13]
```diff
@@ -7,8 +7,33 @@ import (
...
+// Simple logging example using the log package.
+func ExampleSimple() {
+	// UNIX Time is faster and smaller than most timestamps
+	// If you set zerolog.TimeFieldFormat to an empty string,
+	// logs will write with UNIX time
+	zerolog.TimeFieldFormat = ""
+
+	log.Print("hello world")
+
+	// Produces: {"time":1516134303,"level":"debug","message":"hello world"}
```
Comment (rs) [inline log_example_test.go]: Are you going to give the static test time trick a try?

File: README.md (hunk 12 above)
Comment (gilcrest) [reply README.md]: You're totally right - changed and committed!

File: log_example_test.go (hunk 13 above)
Comment (gilcrest) [reply log_example_test.go]: Yep!

Comment (gilcrest) [reply log_example_test.go]: I'll try and figure it out ASAP...  I need to move on to learn the rest of your APIs as well - let me know if this is helpful and I can continue to help fill out some of the documentation.

**Comment by gilcrest** [comment]
Here's the latest - I moved the examples out of log_example_test.go in the zerolog package and into a new file (with the same name) in the `zerolog/log` package since those are the functions I'm truly writing examples for... let me know if you think that makes sense.  

I'm at a bit of a loss now though...  I isolated the time like you recommended and set it to a static time and the function is now providing the correct static output, however, the examples aren't working.  I tried overriding the global logger to ensure the output is being directed to the stdout, but still no luck...  See below - the output is clearly written, but for some reason, the test isn't picking it up (`got:` is blank?).  Everything works fine when I create a logger from the regular `zerolog` package, it's when using the `zerolog/log` package where I hit this issue...

```bash
{"time":1199811905,"level":"debug","message":"hello world"}
--- FAIL: ExamplePrint (0.00s)
got:

want:
{"time":1199811905,"level":"debug","message":"hello world"}
FAIL
```

**Comment by rs** [comment]
There is a strange issue with the testing framework. If you put `log.Logger = zerolog.New(os.Stdout).With().Timestamp().Logger()` at the beginning of each example, it fixes the issue. It does not work if you put it in the init function. I'm not sure why.

**Comment by gilcrest** [comment]
Oh man - thanks for pointing that out, I was pulling my hair out trying to figure out what the hell I was doing wrong.  Submitted another commit, this time with more examples - let me know if this looks ok or if you need anything else.

**Comment by rs** [comment]
@gilcrest do you need some help to complete this PR?

**Comment by gilcrest** [comment]
Hey @rs - I wasn't sure if you're ok with everything at this point?  Do I need to do anything else to finalize the PR?

**Comment by rs** [comment]
Fair enough

**Pull request merged**